packages/client/src/fetchMetadata.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type {
+  ActionDefinition,
+  InterfaceDefinition,
+  ObjectTypeDefinition,
+  QueryDefinition,
+} from "@osdk/api";
+import type {
+  ActionMetadata,
+  InterfaceMetadata,
+  ObjectMetadata,
+  QueryMetadata,
+} from "@osdk/client.api";
+import { OntologiesV2 } from "@osdk/internal.foundry";
+import type { MinimalClient } from "./MinimalClientContext.js";
+import { addUserAgentAndRequestContextHeaders } from "./util/addUserAgentAndRequestContextHeaders.js";
+
+/** @internal */
+export const fetchMetadataInternal = async <
+  Q extends (
+    | ObjectTypeDefinition<any, any>
+    | InterfaceDefinition<any, any>
+    | ActionDefinition<any, any, any>
+    | QueryDefinition<any, any, any>
+  ),
+>(
+  client: MinimalClient,
+  definition: Q,
+): Promise<
+  Q extends ObjectTypeDefinition<any, any> ? ObjectMetadata
+    : Q extends InterfaceDefinition<any, any> ? InterfaceMetadata
+    : Q extends ActionDefinition<any, any, any> ? ActionMetadata
+    : Q extends QueryDefinition<any, any, any> ? QueryMetadata
+    : never
+> => {
+  if (definition.type === "object") {
+    return fetchObjectMetadata(client, definition) as any;
+  } else if (definition.type === "interface") {
+    return fetchInterfaceMetadata(client, definition) as any;
+  } else if (definition.type === "action") {
+    return fetchActionMetadata(client, definition) as any;
+  } else if (definition.type === "query") {
+    return fetchQueryMetadata(client, definition) as any;
+  } else {
+    throw new Error("Type checking should be exhaustive");
```

Comment:
Let's do something more along the lines of `not implemented` , basically something that would be more beneficial for a user